x/withdraw/keeper/ibc_callbacks.go

```diff
@@ -0,0 +1,206 @@
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+
+	"github.com/tharsis/evmos/v2/x/withdraw/types"
+)
+
+// OnRecvPacket performs an IBC receive callback.
+func (k Keeper) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	ack exported.Acknowledgement,
+) exported.Acknowledgement {
+	logger := k.Logger(ctx)
+
+	params := k.GetParams(ctx)
+	claimsParams := k.claimsKeeper.GetParams(ctx)
+
+	// check channels from this chain (i.e destination)
+	if !params.EnableWithdraw || !claimsParams.IsAuthorizedChannel(packet.DestinationChannel) {
+		// return original ACK if withdraw is disabled globally or per channel
+		return ack
+	}
+
+	// unmarshal packet data to obtain the sender and recipient
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		err = sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data")
+		return channeltypes.NewErrorAcknowledgement(err.Error())
+	}
+
+	// validate the sender bech32 address from the counterparty chain
+	bech32Prefix := strings.Split(data.Sender, "1")[0]
+	if bech32Prefix == data.Sender {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender: %s", data.Sender).Error(),
+		)
+	}
+
+	senderBz, err := sdk.GetFromBech32(data.Sender, bech32Prefix)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender %s, %s", data.Sender, err.Error()).Error(),
+		)
+	}
+
+	// change the bech32 human readable prefix (HRP) of the sender to `evmos1`
+	sender := sdk.AccAddress(senderBz)
+
+	// obtain the evmos recipient address
+	recipient, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid receiver address %s", err.Error()).Error(),
+		)
+	}
+
+	if !sender.Equals(recipient) {
+		// return the original success acknowledgement since the address are different
+
+		// NOTE: here we can't know if the recipient is a 'secp256k1' address. If it is,
+		// then the user will need to withdraw the tokens using this same logic.
+		return ack
+	}
+
+	balances := k.bankKeeper.GetAllBalances(ctx, recipient)
+
+	// NOTE: since the balance is 0 we can reject the transaction
+	// We return error since the sender and recipient addresses are the same
+	if balances.IsZero() {
+		logger.Debug(
+			"rejected IBC transfer to 'secp256k1' key address",
+			"sender", data.Sender, "recipient", data.Receiver,
+		)
+
+		// return the an error acknowledgement since the address the same
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(types.ErrKeyTypeNotSupported, "receiver address %s is not a valid ethereum address", data.Receiver).Error(),
+		)
+	}
+
+	// swap the sender and recipient
+	sender = recipient
```

Comment:
Why are you swapping here if the addresses are the same?

Reply:
Also even if they were different this is not a swap. It just replaces the sender